runtime/common/src/main/java/org/apache/nemo/runtime/common/metric/MetricUtils.java

```diff
@@ -0,0 +1,312 @@
+  }
+
+  public static Boolean metaDataLoaded() {
+    return META_DATA_LOADED.getCount() == 0;
+  }
+
+  /**
+   * Save the BiMaps to DB if changes are necessary (rarely executed).
+   */
+  private static void saveBiMaps() {
+    if (!metaDataLoaded()
+      || (MUST_UPDATE_INDEX_EP_BI_MAP.getCount() + MUST_UPDATE_INDEX_EP_KEY_BI_MAP.getCount() == 2)) {
+      // no need to update
+      LOG.info("Not saving Metadata: metadata loaded: {}, Index-EP data: {}, Index-EP Key data: {}",
+        metaDataLoaded(), MUST_UPDATE_INDEX_EP_BI_MAP.getCount() == 0, MUST_UPDATE_INDEX_EP_KEY_BI_MAP.getCount() == 0);
+      return;
+    }
+    LOG.info("Saving Metadata..");
+
+    deregisterBeamDriver();
+    try (final Connection c = DriverManager.getConnection(MetricUtils.POSTGRESQL_METADATA_DB_NAME,
+      "postgres", "fake_password")) {
+      try (final Statement statement = c.createStatement()) {
+        statement.setQueryTimeout(30);  // set timeout to 30 sec.
+
+        if (MUST_UPDATE_INDEX_EP_KEY_BI_MAP.getCount() == 0) {
+          try (final PreparedStatement pstmt = c.prepareStatement(
+            "INSERT INTO " + META_TABLE_NAME + " (key, data) "
+              + "VALUES ('INDEX_EP_KEY', ?) ON CONFLICT (key) DO UPDATE SET data = excluded.data;")) {
+            pstmt.setBinaryStream(1,
+              new ByteArrayInputStream(SerializationUtils.serialize(INDEX_EP_KEY_BI_MAP)));
+            pstmt.executeUpdate();
+            LOG.info("Index-EP Key Metadata saved to DB.");
+          }
+        }
+
+        if (MUST_UPDATE_INDEX_EP_BI_MAP.getCount() == 0) {
+          try (final PreparedStatement pstmt =
+                 c.prepareStatement("INSERT INTO " + META_TABLE_NAME + "(key, data) "
+                     + "VALUES ('INDEX_EP', ?) ON CONFLICT (key) DO UPDATE SET data = excluded.data;")) {
+            pstmt.setBinaryStream(1,
+              new ByteArrayInputStream(SerializationUtils.serialize(INDEX_EP_BI_MAP)));
+            pstmt.executeUpdate();
+            LOG.info("Index-EP Metadata saved to DB.");
+          }
+        }
+      }
+    } catch (SQLException e) {
+      LOG.warn("Saving of Metadata to DB failed: ", e);
+    }
+  }
+
+  /**
+   * Stringify execution properties of an IR DAG.
+   * @param irdag IR DAG to observe.
+   * @return the pair of stringified execution properties. Left is for vertices, right is for edges.
+   */
+  static Pair<String, String> stringifyIRDAGProperties(final IRDAG irdag) {
+    final StringBuilder vStringBuilder = new StringBuilder();
+    final StringBuilder eStringBuilder = new StringBuilder();
+
+    irdag.getVertices().forEach(v ->
+      v.getExecutionProperties().forEachProperties(ep ->
+        epFormatter(vStringBuilder, VERTEX, v.getNumericId(), ep)));
+
+    irdag.getVertices().forEach(v ->
+      irdag.getIncomingEdgesOf(v).forEach(e ->
+        e.getExecutionProperties().forEachProperties(ep ->
+          epFormatter(eStringBuilder, EDGE, e.getNumericId(), ep))));
+
+    saveBiMaps();
```

Comment:
Out of context.

Reply:
Oh it's for saving the metadata, if changes are required. I'll rename the method.